autonomi/src/uploader/upload.rs

```diff
@@ -0,0 +1,1235 @@
+#[cfg(feature = "data")]
+use sn_protocol::{messages::ChunkProof, storage::Chunk};
+use sn_protocol::{storage::RetryStrategy, NetworkAddress};
+#[cfg(feature = "registers")]
+use sn_registers::RegisterAddress;
+use std::{
+    collections::{HashMap, HashSet},
+    num::NonZero,
+};
+use xor_name::XorName;
+
+/// The maximum number of sequential payment failures before aborting the upload process.
+#[cfg(not(test))]
+const MAX_SEQUENTIAL_PAYMENT_FAILS: usize = 3;
+#[cfg(test)]
+const MAX_SEQUENTIAL_PAYMENT_FAILS: usize = 1;
+
+/// The maximum number of sequential network failures before aborting the upload process.
+// todo: use uploader.retry_strategy.get_count() instead.
+#[cfg(not(test))]
+const MAX_SEQUENTIAL_NETWORK_ERRORS: usize = 32;
+#[cfg(test)]
+const MAX_SEQUENTIAL_NETWORK_ERRORS: usize = 1;
+
+/// The number of upload failures for a single data item before
+#[cfg(not(test))]
+const UPLOAD_FAILURES_BEFORE_SELECTING_DIFFERENT_PAYEE: usize = 3;
+#[cfg(test)]
+const UPLOAD_FAILURES_BEFORE_SELECTING_DIFFERENT_PAYEE: usize = 1;
+
+type Result<T> = std::result::Result<T, UploadError>;
+
+// TODO:
+// 1. track each batch with an id
+// 2. create a irrecoverable error type, so we can bail on io/serialization etc.
+// 3. separate cfgs/retries for register/chunk etc
+// 4. log whenever we insert/remove items. i.e., don't ignore values with `let _`
+
+/// The main loop that performs the upload process.
+/// An interface is passed here for easy testing.
+pub(super) async fn start_upload(
+    mut interface: Box<dyn UploaderInterface>,
+) -> Result<UploadSummary> {
+    let mut uploader = interface.take_inner_uploader();
+
+    uploader.validate_upload_cfg()?;
+
+    // Take out the testing task senders if any. This is only set for tests.
+    let (task_result_sender, mut task_result_receiver) =
+        if let Some(channels) = uploader.testing_task_channels.take() {
+            channels
+        } else {
+            // 6 because of the 6 pipelines, 1 for redundancy.
+            mpsc_channel(uploader.cfg.batch_size * 6 + 1)
+        };
+    let (make_payment_sender, make_payment_receiver) = mpsc_channel(uploader.cfg.batch_size);
+
+    uploader.start_payment_processing_thread(
+        make_payment_receiver,
+        task_result_sender.clone(),
+        uploader.cfg.payment_batch_size,
+    )?;
+
+    // chunks can be pushed to pending_get_store_cost directly
+    #[cfg(feature = "data")]
+    {
+        uploader.pending_to_get_store_cost = uploader
+            .all_upload_items
+            .iter()
+            .filter_map(|(xorname, item)| {
+                if let UploadItem::Chunk { .. } = item {
+                    Some((*xorname, GetStoreCostStrategy::Cheapest))
+                } else {
+                    None
+                }
+            })
+            .collect();
+    }
+
+    // registers have to be verified + merged with remote replica, so we have to fetch it first.
+    #[cfg(feature = "registers")]
```

Comment:
so `data` and `registers` features can not be enabled at the same time?
that's bit strange, and is not that user-friendly.
I thought autonomi-cli shall provide different API for upload data or register, but on autonomi level, they shall use the same (or different, if the flow has to be different) functions to complete the task.

Reply:
As discussed in slack, we should be feature gating things here. But maybe we should split this up into smaller components for better segregation in the future.